test/cluster/test_aggregation.py

```diff
@@ -57,25 +60,37 @@ async def test_cancel_mapreduce(manager: ManagerClient):
             s2_mark = await s2_log.mark()
 
             # Prevent finishing local mapreduce tasks on node 2.
-            async with inject_error(manager.api, s2.ip_addr, "mapreduce_pause_dispatch_to_shards"):
-                async def do_select():
-                    # Make node 1 the supercoordinator of the mapreduce task corresponding to aggregation.
-                    # We use this timeout because it's longer than the cumulative timeout of the following
-                    # steps. For the test to be reliable, the query cannot end on its own.
-                    try:
-                        await cql.run_async(f"SELECT count(*) FROM {t} BYPASS CACHE USING TIMEOUT 600s", host=host1)
-                        pytest.fail(f"Query finished, but it wasn't supposed to")
-                    except NoHostAvailable:
-                        pass
-
-                async def wait_and_shutdown():
-                    # Make sure node 1 is the supercoordinator and sends a mapreduce task to node 2.
-                    await s1_log.wait_for(f"dispatching mapreduce_request=.* to address={host_id2}", from_mark=s1_mark, timeout=60)
-                    # Make sure that node 2 is preventing its local mapreduce task from finishing.
-                    await s2_log.wait_for("mapreduce_pause_dispatch_to_shards: waiting for message", from_mark=s2_mark, timeout=60)
-                    # Verify that the supercoordinator stops without an issue despite the ongoing mapreduce task.
+            await manager.api.enable_injection(s2.ip_addr, "mapreduce_pause_dispatch_to_shards", one_shot=False)
+
+            async def do_select():
+                # Make node 1 the supercoordinator of the mapreduce task corresponding to aggregation.
+                # We use this timeout because it's longer than the cumulative timeout of the following
+                # steps. For the test to be reliable, the query cannot end on its own.
+                try:
+                    await cql.run_async(f"SELECT count(*) FROM {t} BYPASS CACHE USING TIMEOUT 600s", host=host1)
+                    pytest.fail(f"Query finished, but it wasn't supposed to")
+                except NoHostAvailable:
+                    pass
+                except OperationTimedOut:
+                    pass
```

Comment:
Is it correct to catch `OperationTimedOut` here? The query is executed with a long timeout and we expect it to be aborted before it times out, right? If so, then if we reach the timeout then it means that the test didn't work because the query was not aborted.

Reply:
The query times out when the test is shutting down 2nd server (the one that responds to `mapreduce_request` RPC). Then node 1 is still doing the query, which then times out.

But I can pass `NoHostAvailable` if `stop_node == sender` and  `OperationTimedOut` if `stop_node == receiver`

Reply:
> Then node 1 is still doing the query, which then times out.

Does it mean that we wait until the full timeout, i.e. 10 minutes? If so then it is definitely too long.

Why is it still doing the query? Is it because of the `mapreduce_pause_dispatch_to_shards` injection? Can we send a message to the injection on the sender node to unblock it?

Reply:
TBH I need to investigate it.

When I think about it, I would expect that `Unavailable` exception is thrown in case `stop_node == receiver`. Because it should go like:
- super-coordinator(node 1) sends a subquery to node 2
- node 2 is shutdown during the execution of the subquery
- the subquery is aborted -> rpc call failed
- retrying dispatcher retries the failed subquery locally but it should also fail (node 2 is already down)
- main aggregation query fails because it cannot gather all data

> Does it mean that we wait until the full timeout, i.e. 10 minutes? If so then it is definitely too long.
When I tested it locally the test didn't take 10 minutes. Again, I need to look why timeout error is thrown.